src/printer.js

```diff
@@ -2330,19 +2337,46 @@ function nodeStr(node, options) {
     shouldUseSingleQuote = true;
   }
 
-  const result = jsesc(str, {
-    quotes: shouldUseSingleQuote ? "single" : "double",
-    wrap: true
-  });
+  const raw = options.parser === 'flow' ? node.raw : node.extra.raw;
+  const isSingleQuote = raw.charAt(0) === "'";
 
-  // Workaround a bug in the Javascript version of the flow parser where
-  // astral unicode characters like \uD801\uDC28 are incorrectly parsed as
-  // a sequence of \uFFFD.
-  if (options.parser === "flow" && result.indexOf("\\uFFFD") !== -1) {
-    return node.raw;
+  if (isSingleQuote === shouldUseSingleQuote) {
+    return raw;
```

Comment:
I'm using `===` as xor. So this is going to be true if both are true or both are false